examples/lockExchange/diagnostics.py

```diff
@@ -0,0 +1,134 @@
+"""
+from thetis import *
+
+
+class FrontLocationCalculator(DiagnosticCallback):
+    """
+    Calculates the location of the propagating front at the top/bottom of the
+    domain.
+
+    This is a useful metric for assessing the accuracy of the model.
+    Theoretical front location speed is
+
+    U = 1/2*sqrt(g*H*delta_rho/rho_0)
+    """
+    name = 'front'
+    variable_names = ['front_bot', 'front_top']
+
+    def _initialize(self, solver_obj):
+        one_2d = Constant(1.0, domain=solver_obj.mesh2d.coordinates.ufl_domain())
+        self.area_2d = assemble(one_2d*dx)
+        # store density difference
+        self.rho = solver_obj.fields.density_3d
+        self.rho_lim = [self.rho.dat.data.min(), self.rho.dat.data.max()]
+        self.delta_rho = self.rho_lim[1] - self.rho_lim[0]
+        self.mean_rho = 0.5*(self.rho_lim[1] + self.rho_lim[0])
+        fs_2d = solver_obj.function_spaces.H_2d
+        mesh2d = fs_2d.mesh()
+        self.xyz = mesh2d.coordinates
+        x = self.xyz.dat.data[:, 0]
+        self.x_lim = [x.min(), x.max()]
+        self.domain_center = 0.5*(self.x_lim[0] + self.x_lim[1])
+        # extractors for surface rho fields
+        self.rho_2d = Function(fs_2d, name='Density 2d')
+        self.extractor_surf = SubFunctionExtractor(self.rho,
+                                                   self.rho_2d,
+                                                   boundary='top',
+                                                   elem_facet='top')
+        self.extractor_bot = SubFunctionExtractor(self.rho,
+                                                  self.rho_2d,
+                                                  boundary='bottom',
+                                                  elem_facet='bottom')
+        # project to higher order cg for interpolation
+        fs_ho = FunctionSpace(mesh2d, 'CG', 2)
+        self.rho_ho = Function(fs_ho, name='Density 2d ho')
+        self.ho_projector = Projector(self.rho_2d, self.rho_ho)
+        self._initialized = True
+
+    def compute_front_location(self):
+        self.ho_projector.project()
+        r = self.rho_ho.dat.data[:]
+        off = 0.25*(self.rho_lim[1] - self.rho_lim[0])
+        up_limit = self.rho_lim[1] - off
+        low_limit = self.rho_lim[0] + off
+        if (r > up_limit).all():  # no front all nodes above limits
+            return self.x_lim[1]
+        if (r < low_limit).all():  # no front all nodes below limits
+            return self.x_lim[0]
+        ix = (r > low_limit) * (r < up_limit)
+        self.rho_ho.dat.data[:] = 0
+        self.rho_ho.dat.data[ix] = 1
+        mass = assemble(self.rho_ho*dx)
+        if mass < 1e-20:
+            return np.nan
+        center_x = assemble(self.rho_ho*self.xyz[0]*dx)/mass
+        return center_x
+
+    def __call__(self, solver_obj):
+        if not hasattr(self, '_initialized') or self._initialized is False:
+            self._initialize(solver_obj)
```

Comment:
Yes, I wanted to avoid storing the solver in the class as a property (solver has a dict of callbacks, who all have solver as a property), but I guess it would be cleaner. In that case `__call__` takes no arguments.